includes/classes/DistributorPost.php

```diff
@@ -0,0 +1,563 @@
+ */
+
+namespace Distributor;
+
+use Distributor\Utils;
+use WP_Post;
+use WP_Term;
+
+/**
+ * This is the post abstraction for distributor posts.
+ *
+ * Although this class is within the Distributor namespace, the class itself
+ * includes the phrase Distributor to make it clear to developers `use`ing
+ * the class that they are not using the `WP_Post` object.
+ *
+ * @since x.x.x
+ */
+class DistributorPost {
+	/**
+	 * The WordPress post object.
+	 *
+	 * @var WP_Post
+	 */
+	public $post = false;
+
+	/**
+	 * Distributable post meta.
+	 *
+	 * An array of
+	 *
+	 * @var array[] {
+	 *    @type mixed[] Post meta keyed by post meta key.
+	 * }
+	 */
+	public $meta = [];
+
+	/**
+	 * Distributable post terms.
+	 *
+	 * @var array[] {
+	 *    @type WP_Term[] Post terms keyed by taxonomy.
+	 * }
+	 */
+	public $terms = [];
+
+	/**
+	 * Distributable post media.
+	 *
+	 * @var array[] Array of media objects.
+	 */
+	public $media = [];
+
+	/**
+	 * Whether this is the source (true) or a distributed post (false).
+	 *
+	 * @var bool
+	 */
+	public $is_source = true;
+
+	/**
+	 * Whether this post is linked to the original version.
+	 *
+	 * For the original post this is set to true.
+	 *
+	 * @var bool
+	 */
+	public $is_linked = true;
+
+	/**
+	 * The original post ID.
+	 *
+	 * @var int
+	 */
+	public $original_post_id = 0;
+
+	/**
+	 * The original post's URL.
+	 *
+	 * This is marked private but can be accessed via the `__get` method to allow
+	 * for live updates of the original post URL for internal connections.
+	 *
+	 * @var string
+	 */
+	private $original_post_url = '';
+
+	/**
+	 * The type of connection this post is distributed from.
+	 *
+	 * @var string internal|external|pushed|empty (for source)
+	 */
+	public $connection_type = '';
+
+	/**
+	 * The connection ID this post is distributed from.
+	 *
+	 * For internal connections this is the site ID. For external connections
+	 * this refers to the connection ID.
+	 *
+	 * @var int
+	 */
+	public $connection_id = 0;
+
+	/**
+	 * The source site data for internal connections.
+	 *
+	 * This is an array of site data for the source site. This is set by
+	 * the populate_source_site() method upon access to avoid switching
+	 * sites unnecessarily.
+	 *
+	 * @var array {
+	 *    @type string $home_url The site's home page.
+	 *    @type string $site_url The site's WordPress address.
+	 *    @type string $rest_url The site's REST API address.
+	 *    @type string $name     The site name.
+	 * }
+	 */
+	private $source_site = [];
+
+	/**
+	 * Initialize the DistributorPost object.
+	 *
+	 * @param WP_Post|int $post WordPress post object or post ID.
+	 */
+	public function __construct( $post ) {
+		$post = get_post( $post );
+
+		if ( ! $post ) {
+			return;
+		}
+
+		$this->post = $post;
+
+		// Set up the distributable data.
+		$this->meta  = Utils\prepare_meta( $post->ID );
+		$this->terms = Utils\prepare_taxonomy_terms( $post->ID );
+		$this->media = $this->prepare_media();
```

Comment:
Wondering if we should set this data in the `constructor` or if it should be set when requested? I guess I'm not sure if this data is needed every time we create an instance of `DistributorPost` or if sometimes this isn't necessary, resulting in potential impacts to performance.

I know this is still a work in progress so maybe this data is always needed but figured I'd call it out

Reply:
> Wondering if we should set this data in the `constructor` or if it should be set when requested?

Good point.

There's an option for limiting which media is distributed so I'll start with that and see if I can either lazy load (similar to source site items) or more simply switch to methods to get all the things.

Reply:
> Wondering if we should set this data in the `constructor` or if it should be set when requested? I guess I'm not sure if this data is needed every time we create an instance of `DistributorPost` or if sometimes this isn't necessary, resulting in potential impacts to performance.

I ended up doing a couple of things:

* storing the media data in a persistent cache for block posts f5d6ddb76c0c94463f75926dfde15c0299a9fe51
* removed the meta, terms and media population from the constractor
* cache them within the object but I may end up removing that to account for post updates.